src/main/java/net/floodlightcontroller/routing/VirtualGatewayInstance.java

```diff
@@ -0,0 +1,303 @@
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import io.netty.util.internal.ConcurrentSet;
+import net.floodlightcontroller.core.types.NodePortTuple;
+import net.floodlightcontroller.routing.web.serializers.VirtualGatewaySerializer;
+import org.projectfloodlight.openflow.types.DatapathId;
+import org.projectfloodlight.openflow.types.IPv4Address;
+import org.projectfloodlight.openflow.types.IPv4AddressWithMask;
+import org.projectfloodlight.openflow.types.MacAddress;
+
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+
+/**
+ * @author Qing Wang (qw@g.clemson.edu) at 12/20/17
+ */
+@JsonSerialize(using = VirtualGatewaySerializer.class)
+@JsonDeserialize(builder = VirtualGatewayInstance.VirtualGatewayInstanceBuilder.class)
+public class VirtualGatewayInstance {
+    protected static final Logger log = LoggerFactory.getLogger(VirtualGatewayInstance.class);
+
+    private final String name;
+    private volatile MacAddress gatewayMac = MacAddress.NONE;
+    private volatile Map<String, VirtualGatewayInterface> interfaces = null;
+
+    private volatile VirtualGatewayInstanceBuilder builder = null;
+    private Set<DatapathId> switchMembers = null;
+    private Set<NodePortTuple> nptMembers = null;
+    private Set<IPv4AddressWithMask> subsetMembers = null;
+
+    public String getName() {
+        return name;
+    }
+    public MacAddress getGatewayMac() {
+        return gatewayMac;
+    }
+    public Collection<VirtualGatewayInterface> getInterfaces() { return interfaces.values(); }
+    public Set<DatapathId> getSwitchMembers() { return switchMembers; }
+    public Set<NodePortTuple> getNptMembers() { return nptMembers; }
+    public Set<IPv4AddressWithMask> getSubsetMembers() { return subsetMembers; }
+    public VirtualGatewayInstanceBuilder getBuilder() { return builder; }
+
+    public boolean isSwitchAMember(DatapathId dpid) { return switchMembers.contains(dpid); }
+    public boolean isNptAMember(NodePortTuple npt) { return nptMembers.contains(npt); }
+    public boolean isSubnetAMember(IPv4AddressWithMask subnet) { return subsetMembers.contains(subnet); }
+
+    public void updateGatewayMac(@Nonnull MacAddress mac) {
+        this.gatewayMac = mac;
+    }
+
+    // add or update interface
+    public void addInterface(VirtualGatewayInterface vInterface) {
+        interfaces.put(vInterface.getInterfaceName(), vInterface);
+    }
+
+    public void addInterface(String name, String ip, String mask) {
+        interfaces.put(name, new VirtualGatewayInterface(name, ip, mask));
+    }
+
+    public Optional<VirtualGatewayInterface> getInterface(String name) {
+        return interfaces.values().stream()
+                .filter(intf -> intf.getInterfaceName().equals(name))
+                .findAny();
+
+    }
+
+    public boolean isInterfaceBelongsGateway(String name) {
```

Comment:
Maybe rename to "interfaceBelongsToGateway"

Reply:
Renamed.